File: src/minilake/mcp/tools/files.py
"""MCP tools for the Files API (/api/2.0/fs) — the modern DBFS replacement."""

from typing import Any

from minilake.mcp.client import MinilakeClient

_PREFIX = "/api/2.0/fs"


def register(mcp: Any, client: MinilakeClient) -> None:
    @mcp.tool()
    async def upload_file(file_path: str, content: str, overwrite: bool = True) -> dict[str, Any]:
        """Upload a text file to minilake's file storage.

        Args:
            file_path: Path without a leading slash, e.g. `data/input.csv`.
            content: File contents as text.
        """
        await client.put(
            f"{_PREFIX}/files/{file_path.lstrip('/')}",
            content=content.encode(),
            params={"overwrite": str(overwrite).lower()},
        )
        return {"path": file_path, "bytes_written": len(content.encode())}

    @mcp.tool()
    async def download_file(file_path: str) -> dict[str, Any]:
        """Download a file's contents as text."""
        raw = await client.get(f"{_PREFIX}/files/{file_path.lstrip('/')}", raw=True)
        return {"path": file_path, "content": raw.decode("utf-8", errors="replace")}

    @mcp.tool()
    async def delete_file(file_path: str) -> dict[str, Any]:
        """Delete a file."""
        await client.delete(f"{_PREFIX}/files/{file_path.lstrip('/')}")
        return {"path": file_path, "deleted": True}

    @mcp.tool()
    async def list_directory(directory_path: str = "") -> dict[str, Any]:
        """List a directory's contents."""
        return await client.get(f"{_PREFIX}/directories/{directory_path.lstrip('/')}")

    @mcp.tool()
    async def create_directory(directory_path: str) -> dict[str, Any]:
        """Create a directory."""
        await client.put(f"{_PREFIX}/directories/{directory_path.lstrip('/')}")
        return {"path": directory_path, "created": True}

File: ui/src/lib/api/files.ts
import { apiFetch, apiFetchRaw } from "./client";
import type { DirectoryEntry } from "./types";

const FS = "/api/2.0/fs";

function encodePath(path: string): string {
  return path
    .replace(/^\/+/, "")
    .split("/")
    .map(encodeURIComponent)
    .join("/");
}

export async function listDirectory(path: string): Promise<DirectoryEntry[]> {
  const data = await apiFetch<{ contents?: DirectoryEntry[] }>(`${FS}/directories/${encodePath(path)}`);
  return data.contents ?? [];
}

export function createDirectory(path: string): Promise<unknown> {
  return apiFetch(`${FS}/directories/${encodePath(path)}`, { method: "PUT" });
}

export function deleteDirectory(path: string): Promise<unknown> {
  return apiFetch(`${FS}/directories/${encodePath(path)}`, { method: "DELETE" });
}

/** Upload raw bytes. The Files API takes the body verbatim, not multipart. */
export function uploadFile(path: string, file: File, overwrite = true): Promise<unknown> {
  return apiFetch(`${FS}/files/${encodePath(path)}`, {
    method: "PUT",
    raw: file,
    query: { overwrite },
    headers: { "Content-Type": "application/octet-stream" },
  });
}

export async function downloadFile(path: string): Promise<Blob> {
  const response = await apiFetchRaw(`${FS}/files/${encodePath(path)}`);
  return response.blob();
}

export function deleteFile(path: string): Promise<unknown> {
  return apiFetch(`${FS}/files/${encodePath(path)}`, { method: "DELETE" });
}
